pkg/util/policy/agent/agent.py

```diff
@@ -0,0 +1,447 @@
+#!/usr/bin/env python
+# Copyright (c) 2016 Pani Networks
+# All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+# http: #www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+from BaseHTTPServer import BaseHTTPRequestHandler,HTTPServer
+from optparse import OptionParser
+import subprocess
+import logging
+import simplejson
+from mimetools import Message
+from StringIO import StringIO
+HTTP_Unprocessable_Entity = 422
+
+logging.basicConfig(level=logging.DEBUG,
+                    format='%(asctime)s %(levelname)-8s %(message)s',
+                    datefmt='%a, %d %b %Y %H:%M:%S')
+
+# TODO should probably discover this from romana root
+addr_scheme = {
+    "network_width" : 8,
+    "host_width" : 8,
+    "tenant_width" : 4,
+    "segment_width" : 4,
+    "endpoint_width" : 8,
+
+    "network_value" : 10,
+}
+
+parser = OptionParser(usage="%prog --port")
+parser.add_option('--port', default=9630, dest="port", type="int",
+                  help="Port number to listen for incoming requests")
+(options, args) = parser.parse_args()
+
+def filter_rules_idx(rules):
+    """
+    Returns 'sweet spot' in iptables rules, index in *filter table where chain
+    definition ends and rules definition begins.
+    """
+    filter_idx = rules.index('*filter')
+    for rule in rules[filter_idx + 1:]:
+        if not rule.startswith(":"):
+            return rules.index(rule) + 1
+
+# We want to receive json object as a POST.
+class AgentHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        self.send_response(200)
+        self.send_header('Content-type','text/html')
+        self.end_headers()
+        self.wfile.write("Romana policy agent")
+        return
+
+    def do_POST(self):
```

Comment:
This is such a long function. Several parts of it could easily be factored out into their own functions.